bare_metal_billing/billing.py

```diff
@@ -0,0 +1,129 @@
+import csv
+import sys
+import math
+from dataclasses import dataclass
+
+from bare_metal_billing import models
+
+
+BM_SU_LIST = [
+    "BM FC430",
+    "BM FC830",
+    "BM GPUA100SXM4",
+    "BM GPUH100",
+]
+
+RESOURCE_CLASS_2_SU_MAPPING = {
+    "lenovo-sd650nv2-a100": "BM GPUA100SXM4",
+    "sd650nv2": "BM GPUA100SXM4",
+    "fc430": "BM FC430",
+    "fc830": "BM FC830",
+}
+
+SU_RESOURCE_LIST = ["vCPUs", "RAM", "vGPUs"]
+
+
+@dataclass
+class InvoiceWriter:
+    HEADERS = [
+        "Invoice Month",
+        "Project - Allocation",
+        "Project - Allocation ID",
+        "Manager (PI)",
+        "Invoice Email",
+        "Invoice Address",
+        "Institution",
+        "Institution - Specific Code",
+        "SU Hours (GBhr or SUhr)",
+        "SU Type",
+        "Rate",
+        "Cost",
+    ]
+
+    invoice_month: str
+    project_invoices: list[models.ProjectUsage]
+    su_rates: models.SURates
+    output_file: str
+
+    def write_csv(self):
+        csv_rows = [self.HEADERS]
+        for project_invoice in self.project_invoices:
+            for su_type, su_hour in project_invoice.su_hours.items():
+                csv_rows.append(
+                    (
+                        self.invoice_month,
+                        project_invoice.project_name,
+                        "ESI Bare Metal",
+                        "",
+                        "",
+                        "",
+                        "",
+                        "",
+                        su_hour,
+                        su_type,
+                        self.su_rates.root[su_type],
+                        self.su_rates.root[su_type] * su_hour,
+                    )
+                )
+
+        with open(self.output_file, "w") as csvfile:
+            csvwriter = csv.writer(csvfile)
+            csvwriter.writerows(csv_rows)
+
+
+def _get_su_type(lease_info: models.BMNodeUsage):
+    lease_resource_class = lease_info.resource_class
+    if lease_resource_class in RESOURCE_CLASS_2_SU_MAPPING:
+        return RESOURCE_CLASS_2_SU_MAPPING[lease_info.resource_class]
+    return models.BM_UNKNOWN_SU
+
+
+def _get_su_quantity(
+    lease_info: models.BMNodeUsage, su_info: models.SUInfoDict, su_type
+):
+    lease_su_info = su_info.root[su_type]
+
+    cpu_count = lease_info.resource_properties.cpu_count
+    ram = lease_info.resource_properties.ram
+    gpu_count = lease_info.resource_properties.gpu_count
+
+    cpu_quantity = 0 if cpu_count == 0 else cpu_count / lease_su_info["vCPUs"]
+    ram_quantity = 0 if ram == 0 else ram / lease_su_info["RAM"]
+    gpu_quantity = 0 if gpu_count == 0 else gpu_count / lease_su_info["vGPUs"]
+
+    su_quantity = math.ceil(max(cpu_quantity, ram_quantity, gpu_quantity))
+    # TODO: What should do we for SUs with 0 SU? Meaning missing cpu and ram data, likely invalid data?
```

Comment:
In the openshift world I went with setting it to 0 SUs of Unkown SU type because lots of pods continue to run without requesting any CPU and/or RAM (technically that means they don't have any guarantees).

In this case, we could do the same i.e. assign 0 SUs of Unkown SU type and log an error and continue generating the report. Or do what you are currently doing and just stop any further processing until we figure out what node had invalid data.

Reply:
@naved001 @knikolla @joachimweyl Following what Naved has done in Openshift billing, if a project has 0 SUs, I will set it as having 0 SUs, and therefore $0 charged as cost. An error will be logged, but the report will still be generated.